### ISSUE TITLE
documentation for install_bazel_dependencies?

### ISSUE THREAD

**Issue by jmhodges** [bazelbuild/rules_nodejs#1118]
I see references to the `install_bazel_dependencies` in some places of the documentation (like in the typescript stuff) and some referencies in e2e tests, but no actual documentation on the doc website for it. The docs don't mention what it does, nor what `load` statement is used to get it.

**Comment by alexeagle** [comment]
Yes, we are missing documentation for the repository rules. I think stardoc knows how to generate this now.

**Comment by alexeagle** [comment]
Oh, thinking about this a bit more - the `install_bazel_dependencies` rule exists in the `@npm` workspace which is generated. So I don't think we can generate any docs for it, we'll just have to add some explanation in prose somewhere.

**Comment by AustinSchuh** [comment]
I struggled mightily as well to find enough docs to get a working example.  I eventually pieced together the following:

```
http_archive(
    name = "build_bazel_rules_nodejs",
    sha256 = "0d9660cf0894f1fe1e9840818553e0080fbce0851169812d77a70bdb9981c946",
    urls = ["https://github.com/bazelbuild/rules_nodejs/releases/download/0.37.0/rules_nodejs-0.37.0.tar.gz"],
)

# Setup the NodeJS toolchain
load("@build_bazel_rules_nodejs//:defs.bzl", "node_repositories", "yarn_install")

node_repositories()

# Setup Bazel managed npm dependencies with the `yarn_install` rule.
#
# To run yarn by hand, use:
#  bazel run @nodejs_linux_amd64//:bin/yarn -- list
# I'm sure there is a better path, but that works...
yarn_install(
    name = "npm",
    package_json = "//:package.json",
    symlink_node_modules = False,
    yarn_lock = "//:yarn.lock",
)

# Install all Bazel dependencies needed for npm packages that supply Bazel rules
load("@npm//:install_bazel_dependencies.bzl", "install_bazel_dependencies")

install_bazel_dependencies()

load("@npm_bazel_typescript//:index.bzl", "ts_setup_workspace")

ts_setup_workspace()
```

**Comment by alexeagle** [comment]
@AustinSchuh  and @jmhodges  Can you help me understand the path you took to get here? I'd expect you to start with `@bazel/create` which sets up a basic WORKSPACE like this without you having to figure anything out, unless you're trying to do something advanced.

**Comment by jmhodges** [comment]
I was adding this to an already existing bazel repository that I've had for a long time now.

**Comment by AustinSchuh** [comment]
This is a bit wordy, but it feels like you are looking for my thought process to help steer future people correctly, so I went for it.

At the risk of hijacking this thread, I've got a large repo with hundreds of targets already in it.  I've been looking for the couple of lines of code I need to get ts_library to work.  Based on experience with other bazel tools, I was looking for something like the above lines.  An http_archive to add, something to import, and likely something to call.  I've been trained (for better or worse) through experience to be careful running tools.  I'm also very cautious about installing tools onto my laptop to manage the build.  I'd much rather figure out how to do the same thing with Bazel, so I don't have to install the tools on everyone's laptops and the tool is versioned.  (Which results in me being cautious about installing yarn and looking for another way)

So I worked backwards.  I found a commit in the commit log for rules_typescript which pointed me to doing everything with yarn_install.  I had found the yarn_install command from a previous attempt.  https://bazelbuild.github.io/rules_nodejs/install.html has some of that.  After a bunch of searching, and some reading of the code, I pieced together that info from above.

From my point of view, if the front page of the docs for rules_nodejs saying to use ```@bazel/create``` had said that it will add the entries to the WORKSPACE file, and also had a good link to a sandboxed version of yarn for future use, I likely wouldn't have gone down this route.  Same goes for the rules_typescript.  The documentation very much felt like how to automagically configure an existing nodejs/typescript project to work with bazel.

**Comment by alexeagle** [comment]
possible dupe of #1131

**Comment by yoonjesung** [comment]
Regarding multiple applications, what's the standard way to install bazel_dependencies in both? Do you just run it twice like:
```
load("@app1_npm//:install_bazel_dependencies.bzl", "install_bazel_dependencies")

install_bazel_dependencies()

load("@app2_npm//:install_bazel_dependencies.bzl", "install_bazel_dependencies")

install_bazel_dependencies()
```

**Comment by sconover** [comment]
Hi, I was following the various setup on instructions on [this site](https://bazelbuild.github.io/rules_nodejs/TypeScript.html), and I needed to include the lines @AustinSchuh mentions:

```
# Install all Bazel dependencies needed for npm packages that supply Bazel rules
load("@npm//:install_bazel_dependencies.bzl", "install_bazel_dependencies")

install_bazel_dependencies()
```

I notice this is mentioned in a couple other places on the web, like [here](https://bazel.googlesource.com/rules_typescript/+/7b471a12c939d3b499f8f00746cf986629df1117/README.md), and am wondering if people should be using [https://bazelbuild.github.io](https://bazelbuild.github.io) given that this documentation seems out of date (and also, for example, the search function is broken).

cc @alexeagle

**Comment by gentunian** [comment]
I can confirm that the generated **npm** workspace while using `yarn` has the generated `install_bazel_dependencies` but it's missing if you are using `npm` instead.

This is the part for `yarn`:
```javascript
function generateInstallBazelDependencies(workspaces) {
    let bzlFile = `# Generated by the yarn_install/npm_install rule
`;
    workspaces.forEach(workspace => {
        bzlFile += `load(\":install_${workspace}.bzl\", \"install_${workspace}\")
`;
    });
    bzlFile += `def install_bazel_dependencies():
    """Installs all workspaces listed in bazelWorkspaces of all npm packages"""
`;
    workspaces.forEach(workspace => {
        bzlFile += `    install_${workspace}()
`;
    });
    writeFileSync('install_bazel_dependencies.bzl', bzlFile);
}
```

being called from:

```javascript
function generateBazelWorkspaces(pkgs) {
    const workspaces = {};
    for (const pkg of pkgs) {
        if (!pkg.bazelWorkspaces) {
            continue;
        }
        for (const workspace of Object.keys(pkg.bazelWorkspaces)) {
            if (workspaces[workspace]) {
                console.error(`Could not setup Bazel workspace ${workspace} requested by npm ` +
                    `package ${pkg._dir}@${pkg.version}. Already setup by ${workspaces[workspace]}`);
                process.exit(1);
            }
            generateBazelWorkspace(pkg, workspace);
            workspaces[workspace] = `${pkg._dir}@${pkg.version}`;
        }
    }
    generateInstallBazelDependencies(Object.keys(workspaces));  // <--- this is missing for npm
}
```

So probably looks like an improvement when you use yarn, but doesn't seem like something important (?), not sure though...
